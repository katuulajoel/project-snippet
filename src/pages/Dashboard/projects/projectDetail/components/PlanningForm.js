/* eslint-disable no-unused-vars */
/* -------------------------------------------------------------------------- */
/*                            External Dependencies                           */
/* -------------------------------------------------------------------------- */
import PropTypes from "prop-types";
import React, { useState } from "react";
import { FormGroup } from "reactstrap";

/* -------------------------- Internel Dependencies ------------------------- */
import Input from "../../../../../components/Input";
import TextArea from "../../../../../components/TextArea";
import FieldError from "../../../../../components/FieldError";
import { validURL } from "../../../../../utils/stringUtils";

/* --------------------------- Style dependencies --------------------------- */
import { StyledForm } from "../../../../../utils/styles";

const PlanningForm = (props) => {
  const { proceed, id } = props;

  const [plan, setPlan] = useState(props.plan || {});
  const [errors, setErrors] = useState({ url: false });

  const validationHandler = (inputName, inputValue) => {
    setErrors({
      ...errors,
      [inputName]:
        inputValue.length >= 1 && !validURL(inputValue) ? "Invalid link" : "",
    });
    return inputValue.length >= 1 && !validURL(inputValue);
  };

  const onChangeValue = (key, value) => {
    let newState = {};
    newState[key] = value;
    setPlan({ ...plan, ...newState });
  };

  const onChangeField = (key, e) => {
    onChangeValue(key, e.target.value);
  };

  const onSave = (e) => {
    e.preventDefault();

    if (validationHandler("url", plan.url)) {
      return;
    }

    if (proceed) {
      proceed(plan);
    }
  };

  return (
    <StyledForm id={id} onSubmit={onSave}>
      <FormGroup>
        {errors.url && <FieldError message="Invalid link" />}
        <Input
          name="url"
          placeholder="Url"
          onChange={(e) => onChangeField("url", e)}
          value={plan.url}
          required
        />
      </FormGroup>
      <FormGroup>
        <Input
          placeholder="Name of file"
          value={plan.title || ""}
          onChange={(e) => onChangeField("title", e)}
          required
        />
      </FormGroup>
      {plan.id ? (
        <FormGroup>
          <label>Reason for changing document</label>
          <TextArea
            value={plan.reason}
            onChange={(e) => onChangeField("reason", e)}
            required
          />
        </FormGroup>
      ) : null}
    </StyledForm>
  );
};

PlanningForm.propTypes = {
  plan: PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    url: PropTypes.string,
    reason: PropTypes.string,
  }),
  proceed: PropTypes.func,
  cancel: PropTypes.func,
  dismiss: PropTypes.func,
  id: PropTypes.string,
};

export default PlanningForm;
