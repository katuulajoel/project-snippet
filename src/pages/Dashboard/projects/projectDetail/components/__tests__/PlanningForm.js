import React from "react";
import renderer from "react-test-renderer";
import PlanningForm from "../PlanningForm";
import { mount } from "enzyme/build";

describe("Project Planning test", () => {
  it("Should match snapshot test", () => {
    const tree = renderer.create(<PlanningForm />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should save form if all required fields are correct", () => {
    const proceedStub = jest.fn();
    const wrapper = mount(
      <PlanningForm plan={{ id: 123 }} proceed={proceedStub} />
    );

    wrapper
      .find("input")
      .first()
      .simulate("change", { target: { value: "url" } });
    wrapper
      .find("input")
      .at(1)
      .simulate("change", { target: { value: "title" } });
    wrapper
      .find("textarea")
      .simulate("change", { target: { value: "description" } });

    wrapper.find("form").simulate("submit");

    expect(proceedStub).toHaveBeenCalledWith({
      id: 123,
      reason: "description",
      url: "url",
      title: "title",
    });
  });
});
