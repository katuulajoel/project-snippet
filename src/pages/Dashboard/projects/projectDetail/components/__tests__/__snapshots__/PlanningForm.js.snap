// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Planning test Should match snapshot test 1`] = `
<form
  className="sc-hKFxyN kdKmpq"
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <div
      className="sc-bdnxRM iEBWwR"
      data-testid="input"
    >
      <input
        className="form-control"
        data-testid=""
        id=""
        label="Label"
        name="url"
        onChange={[Function]}
        placeholder="Url"
        required={true}
        type="input"
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <div
      className="sc-bdnxRM iEBWwR"
      data-testid="input"
    >
      <input
        className="form-control"
        data-testid=""
        id=""
        label="Label"
        onChange={[Function]}
        placeholder="Name of file"
        required={true}
        type="input"
        value=""
      />
    </div>
  </div>
</form>
`;
